### PR TITLE
Module.pthreadPoolSize

### LINKED ISSUE #10231: Add an option to set the pthread pool size at startup
Currently `PTHREAD_POOL_SIZE` is fixed at compile time, but this is of limited use since the running code may know better how many threads are actually needed (say, based on `navigator.hardwareConcurrency`), so we should make it configurable. Perhaps `Module.pthreadPoolSize`?

(see https://twitter.com/prestomation/status/1218276548739944448)

### PR DESCRIPTION
**Pull request by kripken**
This does the same as `PTHREAD_POOL_SIZE` but can be specified
at runtime. While the compile-time setting is great for testing, in
practice I think the Module option is probably more useful, since lots
of projects want to use a pool equal to the number of actual cores
on the user's machine.

fixes #10231

### CODE REVIEW

**Comment by juj** [comment]
Hmm, this is tricky, because it adds un-DCEable code that cannot be optimized away.

How about instead:
 1) use `-s PTHREAD_POOL_SIZE=-1` to mean a compile time decision "give me `navigator.hardwareConcurrency` threads in the pool" (or "give me Module['pthreadPoolSize'] || navigator.hardwareConcurrency` threads in the pool?), and/or
 2) use `-s PTHREAD_POOL_SIZE=2*#+1` to mean a compile time decision "give me 2*navigator.hardwareConcurrency + 1 threads in the pool, and/or
 3) document a code snippet that let's users create an appropriate preRun/--pre-js code that initializes a pool?

That way we would not build more unoptimizable runtime bits. I know the increase in this PR does not add much, but it feels to me that we should strongly prefer to add only DCEable runtime elements.

Or I suppose if the references to `Module['pthreadPoolSize']` are gated behind `-s PTHREAD_POOL_SIZE` setting, then the added size is not bad.

Basically the pthread pool is intended to be an emulation setting for use cases where sync pthread_create() is absolutely required, so I would not want to see all users have to pay bytes for that emulation setting.

**Comment by kripken** [comment]
Fair point about code size. I feel it's a little clunky to need to set an option in order to get `Module.pthreadsPoolSize` to be noticed, but I guess it's not that bad. Changed it that way.

I did consider having special values for "number of cores" or such, but there are going to be more general cases that that doesn't handle, so the Module option seems best.

File: src/preamble.js [hunk 1]
```diff
@@ -780,10 +780,23 @@ var memoryInitializer = null;
...
+#if USE_PTHREADS && PTHREAD_POOL_SIZE
+
+var pthreadPoolSize = Module['pthreadPoolSize'] || {{{ PTHREAD_POOL_SIZE }}};
+
+#if PTHREAD_POOL_DELAY_LOAD != 1
+if (!ENVIRONMENT_IS_PTHREAD && pthreadPoolSize > 0) {
+  addOnPreRun(function() {
+    if (typeof SharedArrayBuffer !== 'undefined') {
+      addRunDependency('pthreads');
+      PThread.allocateUnusedWorkers(pthreadPoolSize, function() {
```
Comment (juj) [inline src/preamble.js]: Uh hmm we have now both `PThread.allocateUnusedWorkers` and `PThread.createNewWorkers(pthreadPoolSize);`, that seems very wrong.  To me this looks like we are allocating 2x the amount of Workers that are needed? Or perhaps I am very confused here. The intent of #9394 where this came into existence does not make sense to me. hmm...

File: src/preamble.js [hunk 2]
```diff
@@ -780,10 +780,23 @@ var memoryInitializer = null;
 #include "memoryprofiler.js"
 #endif
 
-#if PTHREAD_POOL_SIZE > 0 && PTHREAD_POOL_DELAY_LOAD != 1
-// To work around https://bugzilla.mozilla.org/show_bug.cgi?id=1049079, warm up a worker pool before starting up the application.
-if (!ENVIRONMENT_IS_PTHREAD) addOnPreRun(function() { if (typeof SharedArrayBuffer !== 'undefined') { addRunDependency('pthreads'); PThread.allocateUnusedWorkers({{{PTHREAD_POOL_SIZE}}}, function() { removeRunDependency('pthreads'); }); }});
+#if USE_PTHREADS && PTHREAD_POOL_SIZE
+
+var pthreadPoolSize = Module['pthreadPoolSize'] || {{{ PTHREAD_POOL_SIZE }}};
```
Comment (juj) [inline src/preamble.js]: If one wanted to set `Module['pthreadPoolSize'] = 0;` to disable pthread pool, this has the effect of using `{{{PTHREAD_POOL_SIZE}}}` instead?

File: src/library_pthread.js [hunk 3]
```diff
@@ -36,13 +36,14 @@ var LibraryPThread = {
...
+      if (pthreadPoolSize > 0) {
 #if PTHREADS_DEBUG
-      out('Preallocating ' + requestedPoolSize + ' workers.');
-#endif
-      PThread.preallocatedWorkers = PThread.createNewWorkers(requestedPoolSize);
+        out('Preallocating ' + pthreadPoolSize + ' workers.');
 #endif
+        PThread.preallocatedWorkers = PThread.createNewWorkers(pthreadPoolSize);
+      }
+#endif // PTHREAD_POOL_SIZE
```
Comment (juj) [inline src/library_pthread.js]: I am not sure why this code block even exists.. this seems to double allocate a new pool of Workers that has already been created in the code below in `src/preamble.js`?

File: src/preamble.js (hunk 1 above)
Comment (kripken) [reply src/preamble.js]: I believe the idea is that creating and instantiating the workers is a separate thing. One can create the Worker objects without blocking startup on being able to instantiate them with the info they require etc. But we can discuss in your comments over there.

However, all of that is separate from this PR, and should not block it, unless I'm missing something?

Comment (juj) [reply src/preamble.js]: Perhaps this PR can be relayered on top of https://github.com/emscripten-core/emscripten/pull/10269/files#diff-db41bea94577c2dd9b0eef0308b06cf9R36 ?

Comment (kripken) [reply src/preamble.js]: Maybe we can relayer on top of that, but that one seems to need discussion (so we figure out the history there) so unclear how long it will take to land, while this one is ready to go, isn't it? :smile: Or is there some reason you'd prefer to delay landing this?

File: src/preamble.js (hunk 2 above)
Comment (kripken) [reply src/preamble.js]: Oh, good point. Perhaps this should be more careful then.

Another thought I had meanwhile, if we're considering delaying this PR anyhow as discussed above - perhaps `PTHREAD_POOL_SIZE` could be a *string* - then the user could just set it to `navigator.hardwareConcurrency` directly, what do you think?

Comment (juj) [reply src/preamble.js]: I like that idea - they could then set it to `Module['pthreadPoolSize']` as well to customize.

**Comment by kripken** [comment]
Ok, rewritten to simply make `PTHREAD_POOL_SIZE` a string,
PTAL @juj

This makes `PTHREAD_POOL_SIZE` a string, which allows stuff
like `-s PTHREAD_POOL_SIZE=navigator.hardwareConcurrency`
in addition to the old `-s PTHREAD_POOL_SIZE=8` etc. (This works
because our parser is smart and doesn't require unnecessary
quotes for string values.) The value the user provided is simply
emitted in the JS, and used, so it could be anything that is valid
JS.
